src/bun.js/api/bun/x509.zig

```diff
@@ -0,0 +1,553 @@
+        // it from the user either. We use JSON compatible escaping here.
+        _ = BoringSSL.BIO_write(out, "\"", 1);
+        if (safe_prefix) |prefix| {
+            _ = BoringSSL.BIO_printf(out, "%s:", prefix);
+        }
+        for (0..length) |j| {
+            const c = name[j];
+            if (c == '\\') {
+                _ = BoringSSL.BIO_write(out, "\\\\", 2);
+            } else if (c == '"') {
+                _ = BoringSSL.BIO_write(out, "\\\"", 2);
+            } else if ((c >= ' ' and c != ',' and c <= '~') or (utf8 and (c & 0x80) != 0)) {
+                // Note that the above condition explicitly excludes commas, which means
+                // that those are encoded as Unicode escape sequences in the "else"
+                // block. That is not strictly necessary, and Node.js itself would parse
+                // it correctly either way. We only do this to account for third-party
+                // code that might be splitting the string at commas (as Node.js itself
+                // used to do).
+                _ = BoringSSL.BIO_write(out, bun.cast([*]const u8, &c), 1);
+            } else {
+                // Control character or non-ASCII character. We treat everything as
+                // Latin-1, which corresponds to the first 255 Unicode code points.
+                const hex = "0123456789abcdef";
+                const u = [_]u8{ '\\', 'u', '0', '0', hex[(c & 0xf0) >> 4], hex[c & 0x0f] };
+                _ = BoringSSL.BIO_write(out, &u, @sizeOf(@TypeOf(u)));
+            }
+        }
+        _ = BoringSSL.BIO_write(out, "\"", 1);
+    }
+}
+
+inline fn printLatin1AltName(out: *BoringSSL.BIO, name: *BoringSSL.ASN1_IA5STRING, safe_prefix: ?[*]const u8) void {
+    printAltName(out, bun.cast([*]const u8, name.data), @intCast(usize, name.length), false, safe_prefix);
+}
+
+inline fn printUTF8AltName(out: *BoringSSL.BIO, name: *BoringSSL.ASN1_UTF8STRING, safe_prefix: ?[*]const u8) void {
+    printAltName(out, bun.cast([*]const u8, name.data), @intCast(usize, name.length), true, safe_prefix);
+}
+
+pub const kX509NameFlagsRFC2253WithinUtf8JSON = BoringSSL.XN_FLAG_RFC2253 & ~BoringSSL.ASN1_STRFLGS_ESC_MSB & ~BoringSSL.ASN1_STRFLGS_ESC_CTRL;
+
+const NID_ms_upn = 649;
+const NID_id_on_SmtpUTF8Mailbox = 1208;
+const NID_XmppAddr = 1209;
+const NID_SRVName = 1210;
+const NID_NAIRealm = 1211;
+
+// This function emulates the behavior of i2v_GENERAL_NAME in a safer and less
+// ambiguous way. "othername:" entries use the GENERAL_NAME_print format.
+fn x509PrintGeneralName(out: *BoringSSL.BIO, name: *BoringSSL.GENERAL_NAME) bool {
+    if (name.name_type == .GEN_DNS) {
+        _ = BoringSSL.BIO_write(out, "DNS:", 4);
+        // Note that the preferred name syntax (see RFCs 5280 and 1034) with
+        // wildcards is a subset of what we consider "safe", so spec-compliant DNS
+        // names will never need to be escaped.
+        printLatin1AltName(out, name.d.dNSName, null);
```

Comment:
you sure this is always latin1? i feel like we will run into weird bugs in this code; it's safer to write as UTF-8 and then convert if different

Reply:
always latin1, othername can be utf8 and we do the checking, `ASN1_IA5STRING` is always latin1 and `ASN1_UTF8STRING` is always UTF8